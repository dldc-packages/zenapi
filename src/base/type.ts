import type { TEntityAny } from '../entity';
import type { TQuery } from '../query';
import { defineAbstract, defineType } from '../types';

const stringType = defineType<null, undefined>('string');
const dateType = defineType<null, undefined>('date');
const numberType = defineType<null, undefined>('number');
const booleanType = defineType<null, undefined>('boolean');
const jsonType = defineType<null, undefined>('json');
const nilType = defineType<null, undefined>('nil');
const enumType = defineType<readonly string[], undefined>('enum');

export type TNullableDef = { nullable: TQuery | false };

const nullableType = defineType<TEntityAny, TNullableDef>('nullable');

export type TEntityRecord = Record<string, TEntityAny>;

const objectType = defineType<TEntityRecord, string>('object');

export type TListDef =
  | { type: 'all'; select: TQuery }
  | { type: 'first'; select: TQuery }
  | {
      type: 'paginate';
      page: number | { page: number; pageSize: number };
      select: TQuery;
    };

const listType = defineType<TEntityAny, TListDef>('list');

export interface IInputDef {
  input: any;
  select: TQuery;
}

const inputType = defineType<TEntityAny, IInputDef>('input');

export const types = {
  string: stringType,
  date: dateType,
  number: numberType,
  boolean: booleanType,
  json: jsonType,
  nil: nilType,
  enum: enumType,
  nullable: nullableType,
  object: objectType,
  list: listType,
  input: inputType,
};

export type TAbstractObjectDef = Record<string, TQuery>;

const objectAbstract = defineAbstract<TAbstractObjectDef>('object');

const errorBoundaryAbstract = defineAbstract<TQuery>('errorBoundary');

export const abstracts = {
  object: objectAbstract,
  errorBoundary: errorBoundaryAbstract,
};
