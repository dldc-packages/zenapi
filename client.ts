export * from "./src/client/query.ts";
export * from "./src/client/query.types.ts";
export * from "./src/utils/types.ts";
