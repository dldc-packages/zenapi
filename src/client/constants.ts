export const RESULT = Symbol("RESULT");
export const TO_JSON = Symbol("TO_JSON");
