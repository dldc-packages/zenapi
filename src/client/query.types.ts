import type { Primitive } from "../utils/types.ts";
import type { RESULT, TO_JSON } from "./constants.ts";

export type TWithNullable<Base, Result extends TQueryAny> = Base extends null
  ? TQueryBase<Result[typeof RESULT] | null>
  : Result;

export type TQueryAny = TQueryBase<any>;

export type TQueryDef = any[];

export type TVariables = unknown[];

export interface TQueryBase<Result> {
  [RESULT]: Result;
  [TO_JSON]: [TQueryDef, TVariables];
}

export interface TQueryOfObjectSelectFn<Base, ParentNullable extends boolean> {
  <T extends TQueryAny>(
    select: (current: TQueryOf<NonNullable<Base>, false>) => T,
  ): ParentNullable extends true ? TQueryBase<T[typeof RESULT] | null>
    : TQueryBase<T[typeof RESULT]>;
}

export type TQueryOfObject<
  T extends Record<string, any>,
  ParentNullable extends boolean,
> =
  & TQueryBase<null>
  & { [K in keyof T]-?: TQueryOfNested<T[K], ParentNullable> }
  & { _: TQueryOfObjectSelectFn<T, ParentNullable> };

export interface TQueryOfArray<T, ParentNullable extends boolean>
  extends TQueryBase<null> {
  _: TQueryArraySelectFn<T, ParentNullable>;
}

export interface TQueryArraySelectFn<Item, ParentNullable extends boolean> {
  <T extends TQueryAny>(
    select: (current: TQueryOfNested<Item, ParentNullable>) => T,
  ): TQueryBase<T[typeof RESULT][]>;
}

// Check if the type contains null or undefined

// Utility type to check if a type T contains null
type ContainsNull<T> = null extends T ? true : false;
// Utility type to check if a type T contains undefined
type IsUndefined<T> = undefined extends T ? true : false;
// Combined utility type to check if a type T contains null or undefined
type ContainsNil<T> = ContainsNull<T> extends true ? true
  : IsUndefined<T> extends true ? true
  : false;

export type TQueryOfNested<
  T,
  ParentNullable extends boolean,
> = [T] extends [null] ? TQueryBase<null> // Handle null value
  : [T] extends [undefined] ? TQueryBase<null> // Handle undefined value
  : ParentNullable extends true ? TQueryOf<NonNullable<T>, true>
  : ContainsNil<T> extends true ? TQueryOf<NonNullable<T>, true>
  : TQueryOf<T, false>;

export type TQueryOf<T, Nullable extends boolean> = T extends Primitive
  ? TQueryBase<Nullable extends true ? T | null : T>
  : T extends (...args: any) => any
    ? (...args: Parameters<T>) => TQueryOfNested<ReturnType<T>, Nullable>
  : T extends Array<infer U> ? TQueryOfArray<U, Nullable>
  : T extends Record<string, any> ? TQueryOfObject<T, Nullable>
  : never;
