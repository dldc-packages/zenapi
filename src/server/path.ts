import type { TAllStructure } from "./structure.ts";
import type { TGraphRefBase } from "./types.ts";

export type TNormalizedPath = symbol[];

export type TPathCache = WeakMap<
  symbol,
  { ref: TGraphRefBase; structure: TAllStructure }
>;

// export function normalizePath(
//   cache: TPathCache,
//   schema: TSchemaAny,
//   rootStructure: TRootStructure,
//   path: TGraphPath,
// ): TNormalizedPath {
//   const normalized: TNormalizedPath = [];
//   const rootRef = schema.ref as TGraphRefBase;
//   let current: TGraphRefBase = rootRef;
//   let structure: TAllStructure = rootStructure;
//   while (path.length > 0) {
//     const item = path.shift()!;
//     const isRoot = current[GRAPH_KEY] === rootRef[GRAPH_KEY];
//     if (item === ROOT) {
//       if (isRoot) {
//         // Root while already at root, should we just skip?
//         throw new Error(`Invalid path`);
//       }
//       normalized.push(current[GRAPH_KEY]);
//       current = rootRef;
//       structure = rootStructure;
//       continue;
//     }
//     if (structure.kind === "root") {
//       if (!isRoot) {
//         throw new Error(`Invalid path`);
//       }
//       const subStructure: TStructure = structure.types[item];
//       if (!subStructure) {
//         throw new Error(`Invalid path: ${item} not found at root`);
//       }
//       structure = subStructure;
//       current = current[GET](item);
//       continue;
//     }
//     if (isRoot) {
//       throw new Error(`Invalid path`);
//     }
//     if (structure.kind === "object") {
//       const foundProp: TStructureObjectProperty | undefined = structure
//         .properties.find((p) => p.name === item);
//       if (!foundProp) {
//         throw new Error(`Invalid path: ${item} not found in ???`);
//       }
//       structure = foundProp.structure;
//       current = current[GET](item);
//       continue;
//     }
//     if (structure.kind === "ref") {
//       path.unshift(ROOT, structure.ref, item);
//       structure = rootStructure;
//       continue;
//     }
//     throw new Error(`Not implemented ${structure.kind}`);
//   }

//   if (current[GRAPH_KEY] !== rootRef[GRAPH_KEY]) {
//     normalized.push(current[GRAPH_KEY]);
//   }

//   return normalized;
// }
