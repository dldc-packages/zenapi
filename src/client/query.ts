import type { TTypesBase } from "../utils/types.ts";
import { type RESULT, TO_JSON } from "./constants.ts";
import type {
  TQueryAny,
  TQueryBase,
  TQueryDef,
  TQueryOf,
  TQuerySelectFn,
  TVariables,
} from "./query.types.ts";

/**
 * TODO:
 * - Store external data (variables) outside of the query and point to them using a stable ref (hash the path ?)
 * - Make sure to properly handle the _() function
 */

export function query<Types extends TTypesBase>(): TQueryOf<
  Types
> {
  return proxy([], []);
}

type TPath = readonly any[];

function proxy(path: TPath, variables: TVariables): any {
  return new Proxy(
    () => {},
    {
      get(_, prop) {
        if (prop === TO_JSON) {
          return [path, variables];
        }
        if (prop === "_") {
          return ((select) => {
            if (typeof select !== "function") {
              throw new Error("Invalid query function");
            }
            const children = select(proxy([], []));
            const [childrenJson, childVariables] = queryToJson(children);
            return proxy([...path, ...childrenJson], [
              ...variables,
              ...childVariables,
            ]);
          }) satisfies TQuerySelectFn<any>;
        }
        if (prop === "toString") {
          return () => {
            return JSON.stringify(path);
          };
        }
        return proxy([...path, prop], variables);
      },
      apply(_target, _thisArg, parameters) {
        return proxy([...path, "()"], [...variables, parameters]);
      },
    },
  );
}

export function queryToJson(
  query: TQueryAny,
): [query: TQueryDef, variables: TVariables] {
  return query[TO_JSON];
}

function staticQuery<T>(
  def: TQueryDef,
  variables: TVariables,
): TQueryBase<T> {
  return {
    [TO_JSON]: [def, variables],
  } as any;
}

export function obj<T extends Record<string, TQueryAny>>(
  data: T,
): TQueryBase<{ [K in keyof T]: T[K][typeof RESULT] }> {
  const dataJson: { key: string; value: TQueryDef }[] = [];
  const variables: TVariables = [];
  for (const [key, value] of Object.entries(data)) {
    const [itemJson, itemVariables] = queryToJson(value);
    dataJson.push({ key, value: itemJson });
    variables.push(...itemVariables);
  }

  return staticQuery([{ kind: "object", data: dataJson }], variables) as any;
}
