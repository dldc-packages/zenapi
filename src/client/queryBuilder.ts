import type { Primitive, TTypesBase } from "../utils/types.ts";
import { TO_JSON } from "./constants.ts";
import type { TQuery } from "./types.ts";

export type TObjectQueryBuilderFn<T extends Record<string, any>> = <
  Result extends TQuery<any>,
>(
  fnOrQuery:
    | Result
    | ((children: { [K in keyof T]: TQueryBuilderOf<T[K]> }) => Result),
) => Result;

export type TObjectQueryBuilder<T extends Record<string, any>> =
  & { [K in keyof T]: TQueryBuilderOf<T[K]> }
  & { _: TObjectQueryBuilderFn<T> };

export type TQueryBuilderOf<T> = T extends Primitive ? TQuery<T>
  : T extends (...args: any) => any
    ? (...args: Parameters<T>) => TQueryBuilderOf<ReturnType<T>>
  : T extends Array<infer U> ? TQueryBuilderOf<U>
  : T extends Record<string, any> ? TObjectQueryBuilder<T>
  : never;

export function queryBuilder<Types extends TTypesBase>(): TQueryBuilderOf<
  Types
> {
  return proxy("absolute", []);
}

type TPathItem = string | number | symbol | { kind: "call"; args: any[] };

type TPath = readonly TPathItem[];

type TKind = "absolute" | "relative";

function proxy(kind: TKind, path: TPath): any {
  return new Proxy(
    {},
    {
      get(_, prop) {
        if (prop === TO_JSON) {
          return {
            kind,
            path,
          };
        }
        if (prop === "_") {
          return (fnOrQuery: any) => {
            const children = typeof fnOrQuery === "function"
              ? fnOrQuery(proxy("relative", []))
              : fnOrQuery;
            return {
              [TO_JSON]: {
                kind,
                path,
                _: queryToJson(children),
              },
            };
          };
        }
        return proxy(kind, [...path, prop]);
      },
      apply(_target, _thisArg, args) {
        return proxy(kind, [...path, { kind: "call", args }]);
      },
    },
  );
}

export function queryToJson<T>(query: TQuery<T>): unknown {
  const toJsonValue = query[TO_JSON];
  if (toJsonValue) {
    return toJsonValue;
  }
  return query;
}
