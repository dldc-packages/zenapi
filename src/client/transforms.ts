import type { QUERY_RESULT } from "./constants.ts";
import { queryToJson } from "./queryBuilder.ts";
import type { TQuery } from "./types.ts";

export function obj<T extends Record<string, TQuery<any>>>(
  data: T,
): TQuery<{ [K in keyof T]: T[K][typeof QUERY_RESULT] }> {
  return {
    kind: "object",
    data: Object.fromEntries(
      Object.entries(data).map(([key, value]) => [key, queryToJson(value)]),
    ),
  } as any;
}

export function oneOf<T extends readonly TQuery<any>[]>(
  ...queries: T
): TQuery<T[number][typeof QUERY_RESULT]> {
  return { data: queries } as any;
}
