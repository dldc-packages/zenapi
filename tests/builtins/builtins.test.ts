import { assertEquals, assertRejects } from "@std/assert";
import { resolve } from "@std/path";
import { assertSnapshot } from "@std/testing/snapshot";
import * as v from "@valibot/valibot";
import { query, queryToJson } from "../../client.ts";
import {
  builtin,
  createBuiltins,
  createEngine,
  DEFAULT_BUILTINS,
  parse,
  resolver,
  ROOT,
} from "../../server.ts";
import type { MyBuiltin } from "./builtins.ts";
import type { Graph } from "./graph.ts";

interface AllTypes {
  Graph: Graph;
}

const builtins = createBuiltins({
  ...DEFAULT_BUILTINS,
  MyBuiltin: builtin<MyBuiltin>({
    getSchema: () => v.string(),
    match: (value) => typeof value === "string",
    prepare: () => {
      return async (ctx, next) => {
        const res = await next(ctx);
        const value = res.value;
        if (typeof value !== "string") {
          throw new Error("Value is not a string at builtin.MyBuiltin");
        }
        return ctx;
      };
    },
  }),
});

const graph = parse<AllTypes>(
  resolve("./tests/builtins/graph.ts"),
  builtins,
);

Deno.test("Snapshot structure", async (test) => {
  await assertSnapshot(test, graph[ROOT]);
});

const client = query<AllTypes>();

Deno.test("Properly parse MyBuiltin", async () => {
  const engine = createEngine({
    graph,
    entry: "Graph",
    resolvers: [
      resolver(graph.Graph.now, (ctx) => {
        return ctx.withValue("Hello");
      }),
    ],
  });

  const query = client.Graph.now;
  const [queryDef, variables] = queryToJson(query);
  const res = await engine.run(queryDef, variables);
  assertEquals(res, "Hello");
});

Deno.test("Fail if output is not a string", async () => {
  const engine = createEngine({
    graph,
    entry: "Graph",
    resolvers: [
      resolver(graph.Graph.now, (ctx) => {
        return ctx.withValue(42);
      }),
    ],
  });

  const query = client.Graph.now;
  const [queryDef, variables] = queryToJson(query);
  const err = await assertRejects(() => engine.run(queryDef, variables));
  assertEquals(
    (err as Error).message,
    "Value is not a string at builtin.MyBuiltin",
  );
});

Deno.test("MyBuiltin input", async (t) => {
  const engine = createEngine({
    graph,
    entry: "Graph",
    resolvers: [
      resolver(graph.Graph.doStuff.return, (ctx) => {
        const [builtin] = ctx.getInputOrFail(graph.Graph.doStuff);
        return ctx.withValue<string>(builtin);
      }),
    ],
  });

  await t.step("Parse MyBuiltin input", async () => {
    const query = client.Graph.doStuff(
      "Hello",
    );
    const [queryDef, variables] = queryToJson(query);
    const res = await engine.run(queryDef, variables);
    assertEquals(res, "Hello");
  });

  await t.step("Fail if input is not a string", async () => {
    const query = client.Graph.doStuff(42 as any);
    const [queryDef, variables] = queryToJson(query);
    const err = await assertRejects(() => engine.run(queryDef, variables));
    assertEquals(
      (err as Error).message,
      "Invalid arguments passed to root.Graph.doStuff",
    );
  });
});
