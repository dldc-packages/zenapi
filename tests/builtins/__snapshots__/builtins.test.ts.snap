export const snapshot = {};

snapshot[`Snapshot structure 1`] = `
{
  builtins: [
    {
      getSchema: [Function: getSchema],
      key: "builtin.Date",
      kind: "builtin",
      match: [Function: match],
      name: "Date",
      prepare: [Function: prepare],
    },
    {
      getSchema: [Function: getSchema],
      key: "builtin.MyBuiltin",
      kind: "builtin",
      match: [Function: match],
      name: "MyBuiltin",
      prepare: [Function: prepare],
    },
  ],
  key: "root",
  kind: "root",
  mode: "graph",
  types: [
    {
      key: "root.Graph",
      kind: "interface",
      name: "Graph",
      parameters: [],
      properties: [
        {
          name: "now",
          optional: false,
          structure: {
            key: "root.Graph.now",
            kind: "ref",
            params: [],
            ref: "MyBuiltin",
          },
        },
        {
          name: "doStuff",
          optional: false,
          structure: {
            arguments: {
              arguments: [
                {
                  name: "date",
                  optional: false,
                  structure: {
                    key: "root.Graph.doStuff.arguments.date",
                    kind: "ref",
                    params: [],
                    ref: "MyBuiltin",
                  },
                },
              ],
              key: "root.Graph.doStuff.arguments",
              kind: "arguments",
            },
            key: "root.Graph.doStuff",
            kind: "function",
            returns: {
              key: "root.Graph.doStuff.returns",
              kind: "primitive",
              type: "string",
            },
          },
        },
      ],
    },
  ],
}
`;
